### PR TITLE
dns: Support link local IPv6 addresses

### PR DESCRIPTION
**Pull request by arjan-bal**
Fixes: #7882 #3272

Link local IPv6 addresses have the format `[IPv6-address%interface]:port`, notice the extra `%interface` part which is called the `zone`. Unlike link local IPv4 addresses, link local IPv6 addresses are not globally unique, instead they can be re-used for each network interface. As a result, the interface is required to disambiguate link local IPv6 addresses (see [wikipedia](https://en.wikipedia.org/wiki/IPv6_address#Scoped_literal_IPv6_addresses_(with_zone_index))).

Since net.ParseIP doesn't support such addresses, the PR replaces usages of net.ParseIP with the new `ipnet.PareAddr` which support IPv6 addresses with zones and also returns parsing errors.

RELEASE NOTES:
* dns: Add support for link local IPv6 addresses.

### CODE REVIEW

**Review by dfawley — APPROVED** [review]

**Comment by dfawley** [comment]
As mentioned offline:

> Should we audit all uses of `net.ParseIP` and change them (and maybe have `vet.sh` forbid its use)?

File: internal/xds/rbac/matchers.go [hunk 1]
```diff
@@ -344,7 +345,8 @@ func newRemoteIPMatcher(cidrRange *v3corepb.CidrRange) (*remoteIPMatcher, error)
 }
 
 func (sim *remoteIPMatcher) match(data *rpcData) bool {
-	return sim.ipNet.Contains(net.IP(net.ParseIP(data.peerInfo.Addr.String())))
```
Comment (dfawley) [inline internal/xds/rbac/matchers.go:347]: Not sure what this old redundant `net.IP` cast was about, but now it's actually necessary.  :)

**Review by dfawley — APPROVED** [review]
Great, thanks!  Glad to see almost all the other uses were just in tests.

File: scripts/vet.sh [hunk 2]
```diff
@@ -89,6 +89,10 @@ git grep '"github.com/envoyproxy/go-control-plane/envoy' -- '*.go' ':(exclude)*.
 git grep -e 'context.Background()' --or -e 'context.TODO()' -- "*_test.go" | grep -v "benchmark/primitives/context_test.go" | grep -v "credential
 s/google" | grep -v "internal/transport/" | grep -v "xds/internal/" | grep -v "security/advancedtls" | grep -v 'context.WithTimeout(' | not grep -v 'context.WithCancel('
 
+# Disallow usage of net.ParseIP in favour of ipnet.ParseAddr as the former
```
Comment (easwars) [inline scripts/vet.sh:92]: Nit: s/ipnet/netip/

**Review by easwars — APPROVED** [review]

File: scripts/vet.sh (hunk 2 above)
Comment (arjan-bal) [reply scripts/vet.sh:92]: Fixed.

**Pull request merged**